scripts/append-boost-config-macros.cmake

```diff
@@ -4,8 +4,9 @@ cmake_minimum_required(VERSION 3.0)
 set(boost_user_config_file "boost/config/user.hpp")
 
 set(define_without_value_mode FALSE)
+math(EXPR ARGC_COUNT ${CMAKE_ARGC}-1)
 
-foreach(i RANGE ${CMAKE_ARGC})
+foreach(i RANGE ${ARGC_COUNT})
```

Comment:
Actually it's:
* argv[0] = cmake
* argv[1] = -P
* argv[2] = append-boost-config-macros.cmake

So the range should be `foreach(i RANGE 3 ${ARGC_COUNT})`.

Please add it as a comment.

Reply:
> Actually it's:
> - argv[0] = cmake
> - argv[1] = -P
> - argv[2] = append-boost-config-macros.cmake

I'd already noted that but I had doubts that cmake behavior will keep consistent between differente versions or invocation modes. I know that args will be filtered out because there are not prefixed with CONFIG_MACRO.

I checked with a simple cmake scripts and it seems that cmake is very permissive with arguments parsing. For example, non cmake command line arguments before of the ```-P``` are ignored and passed to the scripts:

```
$ cat script.cmake
math(EXPR argc ${CMAKE_ARGC}-1)
foreach(i RANGE ${argc})
  message("ARG${i}: '${CMAKE_ARGV${i}}'")
endforeach()
$ cmake pre1 pre2 -P script.cmake post1 post2
ARG0: 'cmake'
ARG1: 'pre1'
ARG2: 'pre2'
ARG3: '-P'
ARG4: 'cmake-scripts-args.cmake'
ARG5: 'post1'
ARG6: 'post2'
```

> So the range should be foreach(i RANGE 3 ${ARGC_COUNT}).
>
> Please add it as a comment.

Currently this is true when the scripts is called by url_sha_boost* but I prefer a more general comment like:
```
# args without prefix CONFIG_MACRO are filtered out
# (including 'cmake' '-P' 'append-boost-config-macros.cmake') 
````
Are you ok with that?

Reply:
> For example, non cmake command line arguments **before** of the -P are ignored

Actually it's quite a question how it should work because from [documentation](https://cmake.org/cmake/help/latest/manual/cmake.1.html) synopsis is:
```
cmake [(-D <var>=<value>)...] -P <cmake-script-file>
```
Is it designed to pass arguments without `-D`?

I see we are using `-C` by the way:
* https://github.com/ruslo/hunter/blob/b818355f23a8d90541a3843eb867019810235484/cmake/projects/OpenSSL/schemes/url_sha1_openssl.cmake.in#L155-L160

Also I've filed the bug once:: https://gitlab.kitware.com/cmake/cmake/issues/16431 😄 

> args without prefix CONFIG_MACRO are filtered out

I don't think we should rely on that. Such tricks lead to non-trivial bugs.

As a summary I think it's better to refactor this into `cmake -D<var>=<value> -P <cmake-script-file>` and don't use `CMAKE_ARGC`.

Reply:
Please use `foreach(i RANGE 3 ${ARGC_COUNT})` here.